Modular

msabramo opened the pull request:
This changes the UI substantially by making it modular.

Instead of `pydockerize` doing everything, there are subcommands: `write_dockerfiles` and `build`. Probably most commonly, they would be chained together:

``` bash
$ pydockerize -t msabramo/pydockerize -p 2.7,3.3,3.4 write_dockerfiles build
```

but the user also has the freedom to do one without the other.

For example, if the user does not have Docker on their system or otherwise prefers not to run `docker build` locally, they can run just the `write_dockerfiles` phase, commit the resulting `Dockerfile`(s) to version control, and then have something else (a CI server?) run the `build` (`docker build`) step.

I might also in the future add more optional steps, like for example:
- a `write_fig`/`write_compose` step that can generate `fig.yml` files for you.
- a `tag_latest` step that can add the `latest` tag to one of the built images.

Cc: @sudarkoff

msabramo committed:
Split write_dockerfile and invoke_docker_build into stage

in preparation for making them separate commands

msabramo committed:
Make requirements file option rather than arg

msabramo committed:
Can now chain commands

E.g.:

    pydockerize -t msabramo/pydockerize -p 2.7,3.3,3.4 write_dockerfiles build

msabramo committed:
README.rst: Update examples with `write_dockerfiles build`

msabramo committed:
Add help strings

msabramo committed:
Remove commented-out `yield`

msabramo committed:
Allow running `build` without `write_dockerfiles`.

msabramo committed:
README.rst: Update output of --help

msabramo committed:
Use click.[s]echo instead of print

Used click.secho in some places to get some more color.

msabramo committed:
Make sure Dockerfile ends with a newline

msabramo committed:
Simply use `Dockerfile` when only one base_image

This is the common case and what people are accustomed to and allows
`docker build .` to just work.

msabramo committed:
Add `images` command; shows images for a repo

msabramo committed:
Move output from pydockerize to write_dockerfiles

Makes more sense there and doesn't emit a bunch of irrelevant output if
you do -- e.g.: `pydockerize -t gabriel images`

sudarkoff commented on pydockerize.py:
```diff
@@ -84,14 +108,9 @@ def get_cmd_from_procfile(procfile):
-    filename = 'Dockerfile-' + base_image
-    print('write_dockerfile: Writing %s' % filename)
+def write_dockerfile(base_image, requirements_file, filename, cmd, entrypoint):
+    click.echo('write_dockerfile: base_image = %r' % base_image)
```
Only one letter difference between write_dockerfiles and write_dockerfile could be confusing in the output, maybe?

msabramo committed:
Improve output of `build` command

Does a better job telling you what tags were built.

sudarkoff commented:
Looks great! I'm a bit conflicted about the `write_dockerfiles` command name. It's very descriptive, but also kinda verbose. But that's just me and otherwise this looks fantastic!

msabramo committed:
build: On success, do `ctx.invoke(images)`

so that user can see info about built images without having to remember
to add `images` to command line.

msabramo committed:
build: Don't use --file if filename is Dockerfile

This allows greater compatibility with Docker < 1.5

msabramo committed:
When building only 1 image, tag image as "latest"

instead of -- e.g.: "py2.7"

This allows `docker run` to work more simply -- e.g.:

    docker run -it -v $(pwd):/host gabriel

msabramo commented:
Let's rename `write_dockerfiles` to `generate` and then that solves both of the issues that you raised.

msabramo committed:
Rename write_dockerfiles => generate

and write_dockerfile => generate_one

msabramo commented:
> Let's rename write_dockerfiles to generate and then that solves both of the issues that you raised.

Done. Merging...
